src/mongocxx/collection.cpp

```diff
@@ -330,102 +380,102 @@ stdx::optional<result::delete_result> collection::delete_one(
     scoped_bson_t bson_replacement{replacement};
     scoped_bson_t bson_sort{options.sort()};
     scoped_bson_t bson_projection{options.projection()};
 
-    scoped_bson_t reply;
-    reply.flag_init();
+    libmongoc::find_and_modify_opts_set_update(opts, bson_replacement.bson());
 
-    bson_error_t error;
+    if (options.bypass_document_validation()) {
+        libmongoc::find_and_modify_opts_set_bypass_document_validation(
+            opts, *options.bypass_document_validation());
+    }
 
-    options::return_document rd =
-        options.return_document().value_or(options::return_document::k_before);
+    if (options.sort()) {
+        libmongoc::find_and_modify_opts_set_sort(opts, bson_sort.bson());
+    }
 
-    // TODO handle bypass document validation when CDRIVER-945 is done.
-    bool r = mongoc_collection_find_and_modify(
-        _impl->collection_t, bson_filter.bson(), bson_sort.bson(), bson_replacement.bson(),
-        bson_projection.bson(), false, options.upsert().value_or(false),
-        rd == options::return_document::k_after, reply.bson(), &error);
+    if (options.projection()) {
+        libmongoc::find_and_modify_opts_set_fields(opts, bson_projection.bson());
+    }
 
-    if (!r) {
-        throw exception::write(std::move(_impl->gle()), std::make_tuple(error.message, error.code));
+    if (options.upsert().value_or(false)) {
+        flags |= ::MONGOC_FIND_AND_MODIFY_UPSERT;
     }
 
-    bsoncxx::document::view result = reply.view();
+    auto rd = options.return_document().value_or(options::return_document::k_before);
+    if (rd == options::return_document::k_after) {
+        flags |= ::MONGOC_FIND_AND_MODIFY_RETURN_NEW;
+    }
 
-    if (result["value"].type() == bsoncxx::type::k_null)
-        return stdx::optional<bsoncxx::document::value>{};
+    libmongoc::find_and_modify_opts_set_flags(opts,
+                                              static_cast<::mongoc_find_and_modify_flags_t>(flags));
 
-    bsoncxx::builder::stream::document b;
-    b << bsoncxx::builder::stream::concatenate{result["value"].get_document()};
-    return b.extract();
+    return find_and_modify(_impl->collection_t, filter, opts);
 }
 
 stdx::optional<bsoncxx::document::value> collection::find_one_and_update(
     bsoncxx::document::view filter, bsoncxx::document::view update,
     const options::find_one_and_update& options) {
-    scoped_bson_t bson_filter{filter};
+    auto opts = libmongoc::find_and_modify_opts_new();
+    auto opts_cleanup = make_guard([&opts] { libmongoc::find_and_modify_opts_destroy(opts); });
+    int flags = ::MONGOC_FIND_AND_MODIFY_NONE;
     scoped_bson_t bson_update{update};
     scoped_bson_t bson_sort{options.sort()};
     scoped_bson_t bson_projection{options.projection()};
 
-    scoped_bson_t reply;
-    reply.flag_init();
+    libmongoc::find_and_modify_opts_set_update(opts, bson_update.bson());
 
-    bson_error_t error;
+    if (options.bypass_document_validation()) {
```

Comment:
A template local to this file in the unnamed namespace sounds like the least bad way to go about it. Maybe try it and make a call on whether it seems better. Overall, I agree with Samantha on keeping things DRY.